Add ObservedReactor for SwiftUI integration

OhKanghoon opened the pull request:
## Summary
This PR introduces `@ObservedReactor`, a property wrapper that seamlessly bridges ReactorKit's reactive architecture with SwiftUI's declarative UI framework. This allows existing ReactorKit-based projects to adopt SwiftUI views without rewriting their reactor logic.

## Motivation
- SwiftUI requires `ObservableObject` for state management, which is incompatible with ReactorKit's RxSwift-based architecture
- Many iOS apps using ReactorKit want to gradually migrate to SwiftUI without rewriting business logic
- Current solutions require manual bridging code in every SwiftUI view

## Implementation Details

### Core Features
- **`@ObservedReactor` property wrapper**: Automatically observes reactor state changes and triggers SwiftUI view updates
- **`_PrimitiveReactor` protocol**: Provides a clean, explicit API for SwiftUI integration
- **Transparent bridging**: Existing reactors work without modification by conforming to `_PrimitiveReactor`
- **Familiar flow**: Maintains ReactorKit's Action → Mutation → State pattern
- **Explicit API**: Direct access to reactor's action stream and current state

### API Design Philosophy
The API has been designed with **explicitness and simplicity** in mind:

1. **Direct RxSwift integration**: Use `reactor.action.onNext(.action)` to dispatch actions
2. **Explicit state access**: Use `reactor.currentState.property` to read state
3. **Protocol-based**: `_PrimitiveReactor` protocol separates SwiftUI requirements from core ReactorKit

This approach:
- Makes the reactive flow more visible and explicit
- Reduces "magic" and makes code easier to understand
- Aligns better with ReactorKit's core reactive patterns
- Provides a minimal, focused API surface

### Usage Example
```swift
struct CounterView: View {
    @ObservedReactor var reactor = CounterReactor()
    
    var body: some View {
        VStack {
            Text("Count: \(reactor.currentState.count)")
            Button("Increase") {
                reactor.action.onNext(.increase)
            }
            Button("Decrease") {
                reactor.action.onNext(.decrease)
            }
        }
    }
}
```

## Testing
- Comprehensive test suite covering:
  - State observation and updates
  - Action dispatching via `action.onNext()`
  - Memory management and retain cycles
  - SwiftUI integration patterns
- All existing tests pass without modification

## Example Project
Added `SwiftUIExample` demonstrating real-world usage patterns including:
- Basic counter with @ObservedReactor
- Form bindings and two-way data flow
- Alert/toast/dialog patterns
- Shared reactor and dependency injection patterns

<img width="350" alt="Simulator Screenshot - iPhone 16 Pro - 2025-09-27 at 18 41 18" src="https://github.com/user-attachments/assets/68309127-d494-431e-8854-aa90343c5c88" />

## Breaking Changes
None - this is a new feature addition that doesn't affect existing ReactorKit APIs.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - Added a SwiftUI example app showcasing counters, bindings, alerts/toasts/dialogs, and shared-reactor patterns.
  - Added SwiftUI integration for reactive state binding and example components.
- Documentation
  - Added a README for the SwiftUI example.
- Tests
  - Added comprehensive tests for the SwiftUI integration.
- Chores
  - Raised iOS deployment targets to iOS 15.0 across projects.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

OhKanghoon committed:
feat: add ObservedReactor property wrapper for SwiftUI integration

Add ObservedReactor as a property wrapper that bridges ReactorKit with SwiftUI.
This allows existing Reactors designed for UIKit to be used in SwiftUI views
without modification.

Key features:
- Automatic state observation and SwiftUI view updates
- Clean property wrapper syntax for SwiftUI views
- Support for bidirectional binding creation
- Dynamic member lookup for convenient state access
- Minimal interface exposure with fileprivate reactor property

The implementation uses @ObservableObject internally to trigger SwiftUI
updates and provides send() method for action dispatch.

OhKanghoon committed:
test: add comprehensive tests for ObservedReactor

Add test suite covering all ObservedReactor functionality:
- Property wrapper initialization using wrappedValue syntax
- State observation and SwiftUI integration via @ObservableObject
- Action dispatch through send() method
- Bidirectional binding creation with closures and KeyPath
- Dynamic member lookup for state property access
- Memory management and lifecycle verification
- SwiftUI property wrapper usage patterns

Tests verify the wrapper correctly bridges ReactorKit with SwiftUI
and handles all state updates and action dispatching scenarios.

OhKanghoon committed:
feat: add SwiftUIExample demonstrating ObservedReactor usage

Add comprehensive SwiftUI example project showcasing ObservedReactor
integration with ReactorKit. The example demonstrates:

- Counter functionality with increment/decrement actions
- Real-world usage patterns of @ObservedReactor property wrapper
- SwiftUI view binding with ReactorKit state management
- Proper action dispatch and state observation in SwiftUI context

This example serves as both documentation and a working reference
for developers migrating from UIKit to SwiftUI with ReactorKit.

OhKanghoon committed:
build: update iOS deployment target to 15.0 and add SwiftUIExample

Update build configuration to support iOS 15.0 as minimum deployment target:
- Increase platform requirement from iOS 12.0 to 15.0 in Examples Podfile
- Add SwiftUIExample target to Podfile workspace configuration
- Include post_install script to enforce iOS 15.0 deployment target
- Update Podfile.lock with new dependency resolution

This change aligns with SwiftUI requirements and ensures compatibility
with ObservedReactor property wrapper functionality.

OhKanghoon committed:
build: update project configurations for iOS 15.0

Update Xcode project files to reflect new iOS 15.0 deployment target:
- Counter example project configuration updated
- GitHubSearch example project configuration updated

These changes ensure all example projects build correctly with the
new minimum iOS version requirement for SwiftUI compatibility.

ElonPark reviewed (commented):
.

ElonPark reviewed (commented):
Excellent work! 
I have a question what should I do when using `@Pulse` in reactor state?
It would be nice if the example code also included code that uses `@Pulse`.

OhKanghoon commented:
> Excellent work! I have a question what should I do when using `@Pulse` in reactor state? It would be nice if the example code also included code that uses `@Pulse`.

@ElonPark 
Thanks for the question about using `@Pulse`.
I’ve been experimenting with it, and here’s what I found:

### Current Status

**`@Pulse` is not compatible with SwiftUI’s `ObservedReactor`.**
More importantly, I believe it *should not* be made compatible.
The reason is that it fundamentally conflicts with SwiftUI’s declarative philosophy.

### Philosophical Conflict

**UIKit / RxSwift (where `Pulse` makes sense):**

* Imperative: *“When the button is tapped, show an alert.”*
* Event-based: *“An event just happened.”*
* Stream-based: events flow over time

**SwiftUI (where `Pulse` does not fit):**

* Declarative: *“If `showAlert` is true, an alert is visible.”*
* State-driven: *“The UI is determined by current state.”*
* Snapshot-based: one state snapshot defines the whole UI

### The Correct SwiftUI Pattern

#### 1. Alerts & Dialogs → Boolean + content

```swift
struct State {
  var showAlert = false
  var alertTitle = ""
  var alertMessage = ""
}

.alert(alertTitle, isPresented: $reactor.state.showAlert) {
  Button("OK") { $reactor.send(.dismissAlert) }
} message: {
  Text($reactor.state.alertMessage)
}
```

👉 If the same alert is triggered again while it’s already visible, nothing happens — this is **the intended behavior** in SwiftUI, not a limitation.

#### 2. Animations → state-driven

```swift
struct State { var isAnimating = false }

.scaleEffect(isAnimating ? 1.2 : 1.0)
.animation(.spring(), value: isAnimating)
```

#### 3. Navigation → represented as state

```swift
struct State {
  var selectedItem: Item?
  var showingProfile = false
}

.sheet(isPresented: $reactor.state.showingProfile) {
  ProfileView()
}
```

### Why This Is Better

1. **Predictable**: the UI is fully determined by state
2. **Debuggable**: SwiftUI Previews show the exact state
3. **Testable**: just assert on state values
4. **Simple**: no hidden event streams or counters

### Conclusion

**Don’t try to use `@Pulse` in SwiftUI.**
It represents an imperative, event-based mindset that doesn’t align with SwiftUI’s declarative model.

If you’re migrating UIKit code that used `Pulse`, the right approach is:

* Replace it with boolean flags + content in state
* Accept that “same state = no change” is the *correct* behavior
* Trust SwiftUI’s state-driven model

---

That said, this is just my current thinking after experimenting with different approaches.
I’d be very interested to hear other perspectives or ideas on how `@Pulse` could (or should) be used in this context.
If anyone has found patterns that work well in SwiftUI while still leveraging `@Pulse`, I’d love to learn more.

OhKanghoon committed:
build: align iOS deployment target to 15.0 across all example projects

Update IPHONEOS_DEPLOYMENT_TARGET from 15.6 to 15.0 in Counter, GitHubSearch, and SwiftUIExample projects to maintain consistency with the main library configuration.

OhKanghoon committed:
refactor: simplify ObservedReactor API by introducing _PrimitiveReactor protocol

Introduce _PrimitiveReactor protocol to provide a cleaner, more explicit API:
- Remove convenience features from ObservedReactor
- Replace $reactor.send(.action) with reactor.action.onNext(.action)
- Replace $reactor.state with reactor.currentState
- Update tests to reflect new API

This change makes the API more explicit and aligns better with ReactorKit's reactive patterns.

OhKanghoon committed:
refactor: update SwiftUIExample to use new ObservedReactor API

Update all SwiftUI example views to use the new explicit API:
- Use reactor.action.onNext(.action) instead of $reactor.send(.action)
- Use reactor.currentState instead of $reactor.state

This demonstrates the updated ObservedReactor usage patterns.

OhKanghoon committed:
Merge branch 'master' into observed-reactor
